### PR TITLE
[MASWE-0005] Add Sensitive Data Hardcoded in the App Package

### LINKED ISSUE #2543: New Risk - Sensitive Data Hardcoded in the App Package [data-hardcoded-app-package]
## Description

Create a new risk for "Sensitive Data Hardcoded in the App Package (MASVS-STORAGE-1)" using the following information:

Sensitive data (including cryptographic and authentication material) can be hardcoded in the app package, source code, and compiled binary which can be extracted by an attacker.

Create "`risks/MASVS-STORAGE/1-***-****/data-hardcoded-app-package/risk.md`" including the following content:

```yaml
---
title: Sensitive Data Hardcoded in the App Package
alias: data-hardcoded-app-package
platform: [android, ios]
profiles: [L1, L2]
mappings:
- {android: 'https://developer.android.com/privacy-and-security/risks/hardcoded-cryptographic-secrets'}
- masvs-v2: [MASVS-STORAGE-1]

---

## Overview

## Impact

## Modes of Introduction

## Mitigations

```

**To complete the sections follow the guidelines from [Writing MASTG Risks & Tests](https://docs.google.com/document/d/1EMsVdfrDBAu0gmjWAUEs60q-fWaOmDB5oecY9d9pOlg/edit?usp=sharing)**

When creating the corresponding test, use the following areas to guide you:

- app package (APK/IPA)
- app source / compiled binary
- libs

**MASTG v1 Refactoring:**

If the risk has a MASVS v1 ID, you can use it to [search for related tests in the MASTG](https://mas.owasp.org/MASTG/tests/) and use them as input to define your risks and associated tests.
Include the related MASTG v1 tests in the metadata of the risk. For example:

```yaml
mappings:
- mastg-v1: [MASTG-TEST-11]
```

## Acceptance Criteria

- [ ] The risk has been created in the correct directory (`risks/MASVS-STORAGE/1-***-****/data-hardcoded-app-package/risk.md`)
- [ ] The risk content follows the [guidelines](https://docs.google.com/document/d/1EMsVdfrDBAu0gmjWAUEs60q-fWaOmDB5oecY9d9pOlg/edit?usp=sharing)
- [ ] At least one GitHub Issue has been created for the corresponding tests (derived from "Modes of Introduction")
- [ ] The risk indicates the related MASTG v1 tests in its metadata.

### PR DESCRIPTION
**Pull request by juanmanuelmartinez-dekra**
Thank you for submitting a Pull Request to the OWASP MASTG. Please make sure that:

- [ ] Your contribution is written in the 2nd person (e.g. you)
- [ ] Your contribution is written in an active present form for as much as possible.
- [ ] You have made sure that the reference section is up to date (e.g. please add sources you have used, make sure that the references to MITRE/MASVS/etc. are up to date)
- [ ] Your contribution has proper formatted markdown and/or code
- [ ] Any references to website have been formatted as [TEXT](URL “NAME”)
- [ ] You verified/tested the effectiveness of your contribution (e.g.: is the code really an effective remediation? Please verify it works!)

This PR closes #2543.

### CODE REVIEW

**Comment by cpholguera** [comment]
Thanks for opening the PR @juanmanuelmartinez-dekra. Please remember to link PRs to Issues using "Closes #xxxx". In this case, simply add "Closes #2543" to the description.

File: risks/MASVS-STORAGE/1-prevent-hardcoded/data-hardcoded-app-package/risk.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+- In the configuration files or source code for convenience.
+- Within third-party libraries or dependencies that may not follow best security practices.
+
+## Mitigations
+
+To mitigate the risks associated with hardcoded sensitive data, developers should:
+- Use secure storage solutions provided by the platform, such as Android Keystore or iOS Keychain, to store sensitive information.
+- Implement proper key management practices, including key rotation and using environment-specific keys.
+- Avoid storing sensitive data within the app package or source code, and instead retrieve it securely from server-side services.
+- Regularly audit the codebase and dependencies for hardcoded sensitive data.
```
Comment (ZabGo) [inline risks/MASVS-STORAGE/1-prevent-hardcoded/data-hardcoded-app-package/risk.md:35]: Should we give examples of tools they could use for that? For instance [gitLeaks](https://github.com/gitleaks/gitleaks) scans git commit for sensitive data such as password, api keys, etc.

File: risks/MASVS-STORAGE/1-prevent-hardcoded/data-hardcoded-app-package/risk.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+Sensitive data can be hardcoded at various stages of app development, including:
+- During initial development, as a shortcut for testing.
+- In the configuration files or source code for convenience.
+- Within third-party libraries or dependencies that may not follow best security practices.
+
+## Mitigations
+
+To mitigate the risks associated with hardcoded sensitive data, developers should:
+- Use secure storage solutions provided by the platform, such as Android Keystore or iOS Keychain, to store sensitive information.
```
Comment (ZabGo) [inline risks/MASVS-STORAGE/1-prevent-hardcoded/data-hardcoded-app-package/risk.md:32]: Even though the Android Keystore stores keys and certificates, I would not reference it as a storage solution. Storage solutions for Android are Jetpack DataStore and Room, SharedPreferences, etc. It's where the sensitive data will be stored (encrypted of course ;-) ).

File: risks/MASVS-STORAGE/1-prevent-hardcoded/data-hardcoded-app-package/risk.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+- Implement proper key management practices, including key rotation and using environment-specific keys.
+- Avoid storing sensitive data within the app package or source code, and instead retrieve it securely from server-side services.
```
Comment (ZabGo) [inline risks/MASVS-STORAGE/1-prevent-hardcoded/data-hardcoded-app-package/risk.md:34]: Should we give more options to the users as often they will need to store sensitive data locally?

**Review by cpholguera — CHANGES_REQUESTED** [review]

File: risks/MASVS-STORAGE/1-prevent-hardcoded/data-hardcoded-app-package/risk.md (hunk 1 above)
Comment (cpholguera) [inline risks/MASVS-STORAGE/1-prevent-hardcoded/data-hardcoded-app-package/risk.md:35]:
```suggestion
- Regularly audit the codebase and dependencies for hardcoded sensitive data (e.g. using tools such as [gitLeaks](https://github.com/gitleaks/gitleaks)).
- While not foolproof, and **to be used as a last resort** when no other secure options are available, code and resource obfuscation and encryption can deter attackers by making it more difficult to analyze your app and discover hard-coded secrets.
```

File: risks/MASVS-STORAGE/1-prevent-hardcoded/data-hardcoded-app-package/risk.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (cpholguera) [reply risks/MASVS-STORAGE/1-prevent-hardcoded/data-hardcoded-app-package/risk.md:34]: Should we also include something like "consider pre-encrypting the sensitive data and securely retrieving the decryption key from retrieve it from secure server-side services"?

**Comment by cpholguera** [comment]
@ZabGo I've answered to your comments and also added new suggestions, would you mind reviewing please?

**Comment by ZabGo** [comment]
> @ZabGo I've answered to your comments and also added new suggestions, would you mind reviewing please?

@cpholguera so sorry completely missed you comment here. I apologised if it blocked you @juanmanuelmartinez-dekra. I'll review that asap!

**Comment by cpholguera** [comment]
No problem. Thank you, @ZabGo!

File: risks/MASVS-STORAGE/1-prevent-hardcoded/data-hardcoded-app-package/risk.md (hunk 2 above)
Comment (ZabGo) [inline risks/MASVS-STORAGE/1-prevent-hardcoded/data-hardcoded-app-package/risk.md:32]:
```suggestion
- Use platform specific solutions to encrypt sensitive data, such as Android Keystore or iOS Keychain, before storing them.
```

Comment (cpholguera) [reply risks/MASVS-STORAGE/1-prevent-hardcoded/data-hardcoded-app-package/risk.md:32]: what about "Use platform specific solutions to encrypt sensitive data, such as Android Keystore or iOS Keychain."

File: risks/MASVS-STORAGE/1-prevent-hardcoded/data-hardcoded-app-package/risk.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
+title: Sensitive Data Hardcoded in the App Package
```
Comment (cpholguera) [inline risks/MASVS-STORAGE/1-prevent-hardcoded/data-hardcoded-app-package/risk.md:1]:
```suggestion
---
title: Sensitive Data Hardcoded in the App Package
```

File: risks/MASVS-STORAGE/1-prevent-hardcoded/data-hardcoded-app-package/risk.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+- Where possible, generate cryptographic keys dynamically on the device, rather than using predefined keys, and ensure that they don't leave the platform-provided keystore. This approach reduces the risk associated with key transmission and storage.
+- If data must be hardcoded, such as some API keys, be sure to configure them with the minimum required permissions to reduce the impact in case of exposure. Many services allow you to create keys with restricted access, which limits the operations that can be performed.
+- Regularly audit the codebase and dependencies for hardcoded sensitive data (e.g. using tools such as [gitLeaks](https://github.com/gitleaks/gitleaks)).
+- While not foolproof, and **to be used as a last resort** when no other secure options are available, code and resource obfuscation and encryption can deter attackers by making it more difficult to analyze your app and discover hard-coded secrets.
```
Comment (cpholguera) [inline risks/MASVS-STORAGE/1-prevent-hardcoded/data-hardcoded-app-package/risk.md:43]:
```suggestion
- While not foolproof, and **to be used as a last resort** when no other secure options are available, code and resource obfuscation and encryption can deter attackers by making it more difficult to analyze your app and discover hard-coded secrets.

```

**Review by cpholguera — COMMENTED** [review]
@juanmanuelmartinez-dekra the risk file is ready to go. Would you mind continuing now with the rest so we can cover for the original issue? That'd include:

- adding at least one test folder with a test.md file
  - if you want to do Android, that'd be `risks/MASVS-STORAGE/1-prevent-hardcoded/data-hardcoded-app-package/android-data-hardcoded-app-package/test.md`
- add at least one demo (and the corresponding files such as snippers and run.sh)
  -  `risks/MASVS-STORAGE/1-prevent-hardcoded/data-hardcoded-app-package/android-data-hardcoded-app-package/demo-1/demo.md`


In our [guidelines](https://docs.google.com/document/d/1EMsVdfrDBAu0gmjWAUEs60q-fWaOmDB5oecY9d9pOlg/edit#heading=h.ue8tn3i2ff0) you can find links to 3 existing risks that include tests and demos as well as the "how to write" sections. 

You can also see Jan's PR here: https://github.com/OWASP/owasp-mastg/pull/2594/files 

If you have any questions please let me know! Thanks so much!

**Comment by ZabGo** [comment]
> Hi @ZabGo! Can you please let us know when you can continue to work on this or do you have a timeline for it? Thanks

Hi @sushi2k I think the owner of the PR is @juanmanuelmartinez-dekra

**Review by cpholguera — APPROVED** [review]

**Pull request merged**